src/mds/CDentry.cc

```diff
@@ -697,4 +698,25 @@ bool CDentry::scrub(snapid_t next_seq)
   return false;
 }
 
+bool CDentry::check_corruption(bool load, bool abort)
+{
+  auto&& snapclient = dir->mdcache->mds->snapclient;
+  auto next_snap = snapclient->get_last_created()+1;
+  if (first > last || (snapclient->is_server_ready() && first > next_snap)) {
```

Comment:
Will this also be possible that the `first` equals to `last`, which are all `head` just like in https://tracker.ceph.com/issues/38452#note-10 ?

Reply:
Could we also check CInode's corruption in this PR ?

Reply:
If `first > next_snap`, it would trigger when `first == SNAP_HEAD`.

>Could we also check CInode's corruption in this PR ?

It's not a bad idea but writing dedicated tests for that kind of corruption would take time. I believe the corruption is introduced via the dentries so it's best to get a fix in for this ASAP.